File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums){
        int maxProd = nums[0];
        int maxSubArrayProd = nums[0];
        int minSubArrayProd = nums[0];

        for (int ind = 1; ind < nums.length; ind++) {
            int currentMin = minSubArrayProd * nums[ind];
            int currentMax = maxSubArrayProd * nums[ind];

            maxSubArrayProd = Math.max(Math.max(currentMax, currentMin), nums[ind]);
            minSubArrayProd = Math.min(Math.min(currentMax, currentMin), nums[ind]);

            maxProd = Math.max(maxProd, maxSubArrayProd);

        }
        return  maxProd;

    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_product = nums[0]
        max_sub_array_product = nums[0]
        min_sub_array_product = nums[0]

        for indx in range(1, len(nums)):
            current_min = min_sub_array_product * nums[indx]
            current_max = max_sub_array_product * nums[indx]

            max_sub_array_product = max(current_max, current_min, nums[indx])
            min_sub_array_product = min(current_max, current_min, nums[indx])

            max_product = max(max_product , max_sub_array_product, min_sub_array_product, nums[indx])

        return max_product